# 📊 Excel Analysis Report

**File:** `Business Accounting.xlsx`


## 🛡️ Security Analysis

**Risk Level:** 🟡 MEDIUM

### Detected Threats:
- **DANGEROUS_FORMULA**: Potentially dangerous function HYPERLINK detected


## 📋 Structural Analysis

**Total Sheets:** 10
**Total Cells with Data:** 6,959
**Named Ranges:** 0
**Complexity Score:** 32/100

### Sheet Details:

| Sheet Name | Hidden | Rows | Columns | Formulas |
|------------|--------|------|---------|----------|
| Yiriden Transactions 2025 | False | 42 | 9 | 2 |
| Yiriden Transactions 2023 | True | 1,021 | 11 | 815 |
| Yiriden 2023 Loans | True | 122 | 14 | 0 |
| Sanoun Transactions 2024 | True | 255 | 28 | 2 |
| Sanoun Transactions 2025 | True | 264 | 11 | 2 |
| 2024 Shea butter shipping | True | 1,001 | 5 | 13 |
| Yiriden mileages | True | 2 | 3 | 0 |
| Truck Revenue Projections | True | 39 | 9 | 39 |
| Yiriden 2022 | True | 1,000 | 26 | 3 |
| Real Estate - Horton Rd | False | 1,001 | 26 | 1 |


## 🔗 Formula Analysis

**Max Dependency Depth:** 0 levels
**Formula Complexity Score:** 1837.01/100
**Circular References:** ✅ No
**Volatile Formulas:** 0
**External References:** 0


In [1]:
import pandas as pd
from pathlib import Path

# Load the Excel file (path relative to repo root)
excel_path = Path("test_assets/collection/business-accounting/Business Accounting.xlsx")
sheet_index = 2

print(f"Loading data from: {excel_path}")
print(f"Loading sheet at index: {sheet_index}")

try:
    # First, get information about all sheets
    xl_file = pd.ExcelFile(excel_path)
    sheet_names = xl_file.sheet_names
    print(f"\nAvailable sheets: {sheet_names}")

    if sheet_index >= len(sheet_names):
        print(f"\nError: Sheet index {sheet_index} is out of range. File has {len(sheet_names)} sheets.")
        print("Please use --sheet-index with a value between 0 and {}".format(len(sheet_names)-1))
        df = None
    else:
        # Load the specified sheet
        selected_sheet = sheet_names[sheet_index]
        print(f"\nLoading sheet: '{selected_sheet}'")

        df = pd.read_excel(excel_path, sheet_name=sheet_index)
        print(f"\nLoaded data with shape: {df.shape}")
        print(f"Columns: {list(df.columns)}")

        # Display first few rows
        print("\nFirst 5 rows:")
        display(df.head())

        # Basic info
        print("\nData types:")
        print(df.dtypes)

except Exception as e:
    print(f"Error loading Excel file: {e}")
    df = None


Loading data from: test_assets/collection/business-accounting/Business Accounting.xlsx
Loading sheet at index: 2

Available sheets: ['Yiriden Transactions 2025', 'Yiriden Transactions 2023', 'Yiriden 2023 Loans', 'Sanoun Transactions 2024', 'Sanoun Transactions 2025', '2024 Shea butter shipping', 'Yiriden mileages', 'Truck Revenue Projections', 'Yiriden 2022', 'Real Estate - Horton Rd']

Loading sheet: 'Yiriden 2023 Loans'

Loaded data with shape: (121, 14)
Columns: ['Lender', 'Amount', 'Term', 'Rate', 'Payment', 'Origination Date', 'Truck', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13']

First 5 rows:


Lender   Amount   Term    Rate  Payment Origination Date  \
0                   BHG  89395.0  120.0  0.1824  1624.60       2023-04-18   
1  Fidelity 401(k) loan  36000.0   59.0  0.0925   350.63       2023-07-16   
2                   NaN      NaN    NaN     NaN      NaN              NaT   
3                   NaN      NaN    NaN     NaN      NaN              NaT   
4                   NaN      NaN    NaN     NaN      NaN              NaT   

      Truck  Unnamed: 7 Unnamed: 8           Unnamed: 9 Unnamed: 10  \
0  Actros 1         NaN  Payment #                 Date      Amount   
1  Actros 2         NaN          1  2023-05-20 00:00:00      1624.6   
2       NaN         NaN          2  2023-06-20 00:00:00      1624.6   
3       NaN         NaN          3  2023-07-20 00:00:00      1624.6   
4       NaN         NaN          4  2023-08-20 00:00:00      1624.6   

  Unnamed: 11 Unnamed: 12 Unnamed: 13  
0    Interest   Principal     Balance  
1         NaN         NaN         NaN  
2         NaN         NaN         NaN  
3         NaN         NaN         NaN  
4         NaN         NaN         NaN


Data types:
Lender                      object
Amount                     float64
Term                       float64
Rate                       float64
Payment                    float64
Origination Date    datetime64[ns]
Truck                       object
Unnamed: 7                 float64
Unnamed: 8                  object
Unnamed: 9                  object
Unnamed: 10                 object
Unnamed: 11                 object
Unnamed: 12                 object
Unnamed: 13                 object
dtype: object


In [2]:
# Query interface for formula dependency analysis
# The pipeline has already analyzed all formulas and cached the results

import pickle
from pathlib import Path
from spreadsheet_analyzer.graph_db.query_interface import create_enhanced_query_interface

# Load cached formula analysis
cache_file = Path(r".pipeline_cache/Business Accounting_formula_analysis.pkl")
with open(cache_file, 'rb') as f:
    formula_analysis = pickle.load(f)

query_interface = create_enhanced_query_interface(formula_analysis)

# Convenience functions for graph queries
def get_cell_dependencies(sheet, cell_ref):
    """Get complete dependency information for a specific cell."""
    result = query_interface.get_cell_dependencies(sheet, cell_ref)
    print(f"\nCell {sheet}!{cell_ref}:")
    print(f"  Has formula: {result.has_formula}")
    if result.formula:
        print(f"  Formula: {result.formula}")
    if result.direct_dependencies:
        print(f"  Direct dependencies: {', '.join(result.direct_dependencies[:5])}")
        if len(result.direct_dependencies) > 5:
            print(f"    ...and {len(result.direct_dependencies) - 5} more")
    if result.direct_dependents:
        print(f"  Cells that depend on this: {', '.join(result.direct_dependents[:5])}")
        if len(result.direct_dependents) > 5:
            print(f"    ...and {len(result.direct_dependents) - 5} more")
    return result

def find_cells_affecting_range(sheet, start_cell, end_cell):
    """Find all cells that affect any cell within the specified range."""
    result = query_interface.find_cells_affecting_range(sheet, start_cell, end_cell)
    print(f"\nCells affecting range {sheet}!{start_cell}:{end_cell}:")
    for cell, deps in list(result.items())[:5]:
        print(f"  {cell} depends on: {', '.join(deps[:3])}")
        if len(deps) > 3:
            print(f"    ...and {len(deps) - 3} more")
    if len(result) > 5:
        print(f"  ...and {len(result) - 5} more cells")
    return result

def get_formula_statistics():
    """Get comprehensive statistics about formulas in the workbook."""
    stats = query_interface.get_formula_statistics_with_ranges()
    print("\nFormula Statistics:")
    print(f"  Total formulas: {stats['total_formulas']:,}")
    print(f"  Formulas with dependencies: {stats['formulas_with_dependencies']:,}")
    print(f"  Unique cells referenced: {stats['unique_cells_referenced']:,}")
    print(f"  Max dependency depth: {stats['max_dependency_depth']} levels")
    print(f"  Circular references: {stats['circular_reference_chains']}")
    print(f"  Formula complexity score: {stats['complexity_score']}/100")
    return stats

def find_empty_cells_in_formula_ranges(sheet):
    """Find empty cells that are part of formula ranges."""
    result = query_interface.find_empty_cells_in_formula_ranges(sheet)
    print(f"\nEmpty cells in formula ranges for sheet '{sheet}':")
    if result:
        print(f"  Found {len(result)} empty cells")
        # Group by rows for display
        rows = {}
        for cell in list(result)[:20]:
            row_num = ''.join(filter(str.isdigit, cell))
            if row_num not in rows:
                rows[row_num] = []
            rows[row_num].append(cell)
        for row, cells in list(rows.items())[:5]:
            print(f"  Row {row}: {', '.join(cells)}")
        if len(result) > 20:
            print(f"  ...and {len(result) - 20} more")
    else:
        print("  No empty cells found in formula ranges")
    return result


## 🔍 Formula Analysis Tools

You have TWO approaches available for formula analysis:

### 1️⃣ Graph-Based Dependency Analysis (Recommended for Complex Files)
The deterministic pipeline has analyzed all formulas and created a dependency graph. These tools are robust and handle complex Excel files:

- **get_cell_dependencies** - Analyze what a cell depends on and what depends on it
- **find_cells_affecting_range** - Find all cells that affect a specific range
- **find_empty_cells_in_formula_ranges** - Find gaps in data that formulas reference
- **get_formula_statistics** - Get overall statistics about formulas
- **find_circular_references** - Find all circular reference chains

### 2️⃣ Formulas Library for Advanced Formula Evaluation (Recommended)
Robust formula evaluation using the 'formulas' library that handles complex Excel files:

- **load_excel_with_formulas** - Load Excel file for formula evaluation
- **evaluate_cell** - Get calculated cell values and formulas
- **set_cell_and_recalculate** - What-if analysis with recalculation
- **get_cell_dependencies_formulas** - Track formula dependencies
- **export_formulas_model** - Export model to JSON
- **get_formulas_help** - Get detailed help

✅ **Recommended**: The formulas library handles complex Excel files much better than other alternatives.

### Usage:
All tools are available through the tool-calling interface. Use graph-based analysis for quick dependency queries, and the formulas library for accurate formula evaluation and what-if analysis.


## --- LLM Analysis Start ---

In [3]:
# Multi-table detection analysis
import numpy as np

# 1. Check for empty row patterns
empty_rows = df.isnull().all(axis=1)
empty_row_groups = empty_rows.groupby((~empty_rows).cumsum()).sum()
print("\nEmpty row blocks:")
print(empty_row_groups[empty_row_groups > 0].to_dict())

# 2. Check for empty columns
empty_cols = df.isnull().all(axis=0)
print("\nEmpty columns:")
print(list(df.columns[empty_cols]))

# 3. Analyze data distribution
print("\nValue distribution in first 30 rows:")
non_null_counts = df.head(30).notna().sum()
print(non_null_counts)

# 4. Check for semantic shifts - analyze value patterns
print("\nUnique values per column (first 7 columns):")
for col in df.columns[:7]:  # Main columns only
    unique_count = df[col].nunique()
    print(f"{col}: {unique_count} unique values")


Empty row blocks:
{}

Empty columns:
['Unnamed: 7']

Value distribution in first 30 rows:
Lender               2
Amount               2
Term                 2
Rate                 2
Payment              2
Origination Date     2
Truck                2
Unnamed: 7           0
Unnamed: 8          30
Unnamed: 9          30
Unnamed: 10         30
Unnamed: 11          1
Unnamed: 12          1
Unnamed: 13          1
dtype: int64

Unique values per column (first 7 columns):
Lender: 2 unique values
Amount: 2 unique values
Term: 2 unique values
Rate: 2 unique values
Payment: 2 unique values
Origination Date: 2 unique values
Truck: 2 unique values


In [4]:
# Comprehensive data analysis
import pandas as pd
pd.set_option('display.max_columns', None)

# 1. Basic statistics for numeric columns
print("Numeric Column Statistics:")
print(df[['Amount', 'Term', 'Rate', 'Payment']].describe())

# 2. Missing value analysis
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)
print("\nMissing Value Analysis:")
for col, miss in zip(df.columns, missing_pct):
    print(f"{col}: {miss}%")

# 3. Loan amount distribution
print("\nLoan Amount Distribution:")
amount_quartiles = df['Amount'].quantile([0.25, 0.5, 0.75])
print(amount_quartiles)

# 4. Payment analysis
print("\nPayment Statistics:")
payment_stats = df.groupby('Lender')['Payment'].agg(['count', 'mean', 'sum']).round(2)
print(payment_stats)

# 5. Term analysis
print("\nLoan Term Distribution:")
term_dist = df['Term'].value_counts().sort_index()
print(term_dist)

# 6. Interest Rate Analysis
print("\nInterest Rate Statistics:")
rate_stats = df['Rate'].describe()
print(rate_stats)

Numeric Column Statistics:
             Amount        Term      Rate      Payment
count      2.000000    2.000000  2.000000     2.000000
mean   62697.500000   89.500000  0.137450   987.615000
std    37755.966581   43.133514  0.063569   900.832826
min    36000.000000   59.000000  0.092500   350.630000
25%    49348.750000   74.250000  0.114975   669.122500
50%    62697.500000   89.500000  0.137450   987.615000
75%    76046.250000  104.750000  0.159925  1306.107500
max    89395.000000  120.000000  0.182400  1624.600000

Missing Value Analysis:
Lender: 98.35%
Amount: 98.35%
Term: 98.35%
Rate: 98.35%
Payment: 98.35%
Origination Date: 98.35%
Truck: 98.35%
Unnamed: 7: 100.0%
Unnamed: 8: 0.0%
Unnamed: 9: 0.0%
Unnamed: 10: 0.0%
Unnamed: 11: 99.17%
Unnamed: 12: 99.17%
Unnamed: 13: 99.17%

Loan Amount Distribution:
0.25    49348.75
0.50    62697.50
0.75    76046.25
Name: Amount, dtype: float64

Payment Statistics:
                      count     mean      sum
Lender                               

In [5]:
# Temporal and relationship analysis

# 1. Loan origination timeline
print("Loan Origination Timeline:")
timeline = df.groupby(df['Origination Date'].dt.to_period('M'))['Amount'].agg(['count', 'sum'])
print(timeline)

# 2. Rate vs Amount correlation
print("\nCorrelation between Rate and Amount:")
corr = df[['Rate', 'Amount', 'Term', 'Payment']].corr().round(3)
print(corr)

# 3. Check for data anomalies
print("\nPotential Anomalies:")
# Check for unusually high/low rates
rate_bounds = df['Rate'].quantile([0.01, 0.99])
print("Rate bounds (1st-99th percentile):", rate_bounds)

# Check payment to amount ratio
df['payment_ratio'] = df['Payment'] / df['Amount']
print("\nPayment to Amount Ratio Statistics:")
print(df['payment_ratio'].describe())

Loan Origination Timeline:
                  count      sum
Origination Date                
2023-04               1  89395.0
2023-07               1  36000.0

Correlation between Rate and Amount:
         Rate  Amount  Term  Payment
Rate      1.0     1.0   1.0      1.0
Amount    1.0     1.0   1.0      1.0
Term      1.0     1.0   1.0      1.0
Payment   1.0     1.0   1.0      1.0

Potential Anomalies:
Rate bounds (1st-99th percentile): 0.01    0.093399
0.99    0.181501
Name: Rate, dtype: float64

Payment to Amount Ratio Statistics:
count    2.000000
mean     0.013956
std      0.005963
min      0.009740
25%      0.011848
50%      0.013956
75%      0.016065
max      0.018173
Name: payment_ratio, dtype: float64


# 📊 Analysis Complete

## Executive Summary
Analysis of the "Yiriden 2023 Loans" sheet reveals a structured loan tracking system with significant data quality issues, including 98.35% missing values in core columns. The sheet contains loan details for vehicle financing with varying terms, rates, and payment structures.

## Data Overview
- Dataset: 121 rows × 14 columns
- Core metrics: Lender, Amount, Term, Rate, Payment, Origination Date, Truck
- Only 2 complete loan records (98.35% missing data)
- Multiple unnamed columns present

## Key Findings
1. **Data Completeness Issues**
   - Only 2 out of 121 rows contain loan data
   - Core columns show 98.35% missing values
   - Unnamed columns 8-10 show full data population

2. **Loan Characteristics**
   - Loan amounts range: $36,000 to $89,395
   - Terms range: 59 to 120 months
   - Interest rates: 9.25% to 18.24%
   - Monthly payments: $350.63 to $1,624.60

3. **Temporal Pattern**
   - Two loans originated in 2023:
     - April 2023: $89,395 loan
     - July 2023: $36,000 loan
   - Suggests sporadic loan activity

4. **Financial Metrics**
   - Payment to Amount ratio: 0.97% to 1.82%
   - Perfect correlation between metrics suggests potential formula relationships
   - Higher loan amounts associated with longer terms

## Data Quality Issues
- Massive data gaps (98.35% missing values)
- Unnamed columns present
- Empty column (Unnamed: 7)
- Possible hidden or filtered data

## Statistical Insights
- Mean loan amount: $62,697.50
- Average term: 89.5 months
- Mean interest rate: 13.75%
- Average payment: $987.62

## Business Implications
- High missing data suggests potential record keeping issues
- Wide range of loan terms (59-120 months) indicates flexible financing options
- Interest rate spread (9.25%-18.24%) suggests risk-based pricing
- Vehicle financing appears to be a significant business activity

## Recommendations
1. **Immediate Actions**
   - Investigate missing data (98.35% gap)
   - Review data entry processes
   - Standardize column naming

2. **Short-term Improvements**
   - Implement data validation rules
   - Create standardized loan tracking template
   - Remove unused columns

3. **Long-term Considerations**
   - Develop proper database for loan tracking
   - Implement automated data quality checks
   - Create loan performance monitoring system

## Technical Notes
- Analysis performed using pandas DataFrame operations
- Missing value analysis conducted across all columns
- Statistical analysis focused on complete records
- Correlation analysis limited by data availability